src/main/java/com/amazon/ion/impl/macro/MacroCompiler.kt

```diff
@@ -0,0 +1,240 @@
+        reader.readContainer {
+            confirm(reader.next() == IonType.SYMBOL && reader.stringValue() == "macro") { "macro compilation expects a sexp starting with the keyword `macro`" }
+
+            nextAndCheckType(IonType.SYMBOL, "macro name")
+            confirmNoAnnotations("macro name")
+            // TODO: Enforce 'identifier' syntax subset of symbol
+            //       Possibly add support for macro definitions without names?
+            macroName = symbolValue().assumeText()
+
+            nextAndCheckType(IonType.SEXP, "macro signature")
+            confirmNoAnnotations("macro signature")
+            readSignature()
+
+            forEachRemaining { compileTemplateBodyExpression(isQuoted = false) }
+        }
+        return TemplateMacro(signature.toList(), expressions.toList())
+    }
+
+    /**
+     * Reads the macro signature, populating parameters in [signature].
+     * Caller is responsible for making sure that the reader is positioned on (but not stepped into) the signature sexp.
+     */
+    private fun readSignature() {
+        reader.forEachInContainer {
+            when (it) {
+                IonType.SYMBOL -> addParameter(grouped = false)
+                IonType.LIST -> {
+                    confirmNoAnnotations(location = "grouped parameter enclosing list")
+                    readContainer {
+                        nextAndCheckType(IonType.SYMBOL, "parameter name")
+                        addParameter(grouped = true)
+                        confirm(next() == null) { "grouped parameter list must enclose only one variable name" }
+                    }
+                }
+                else -> throw IonException("parameter must be a symbol or a list; found ${reader.type}")
+            }
+        }
+    }
+
+    /**
+     * Adds a parameter to the macro signature.
+     * Caller is responsible for making sure that the reader is positioned on a parameter name.
+     */
+    private fun addParameter(grouped: Boolean) {
+        val annotations = reader.typeAnnotations
+        confirm(annotations.isEmptyOr(Macro.ParameterEncoding.Tagged.ionTextName)) { "unsupported parameter encoding ${annotations.toList()}" }
+        val parameterName = reader.symbolValue().assumeText()
+        confirm(signature.none { it.variableName == parameterName }) { "redeclaration of parameter '$parameterName'" }
+        signature.add(Macro.Parameter(parameterName, Macro.ParameterEncoding.Tagged, grouped))
+    }
+
+    /**
+     * Compiles the current value on the reader into a [TemplateBodyExpression] and adds it to [expressions].
+     * Caller is responsible for ensuring that the reader is positioned on a value.
+     *
+     * If called when the reader is not positioned on any value, throws [IllegalStateException].
+     */
+    private fun compileTemplateBodyExpression(isQuoted: Boolean) {
+        // TODO: Could typeAnnotations ever be an array of nulls?
```

Comment:
Nope. The array will be empty if there are no annotations. If any annotation has unknown text, a SymbolToken will still be present in the array (it will just have null text).